### PR TITLE
ras_ppcutils.py: Validating lparstat functionality

### PR DESCRIPTION
**Pull request by shirishaganta1**
Added a new testcase to validate below two
1.laprstat -E output %busy and %idle should not be < 0 and > 100 
2.Normalized %busy + %idle = percentage under frequency

### CODE REVIEW

File: ras/ras_ppcutils.py [hunk 1]
```diff
@@ -447,3 +449,40 @@ def test_lparnumascore(self):
...
+        should not be < 0 or > 100
+        Normalized %busy + %idle should be equal to percentage under frequency
+        """
+        output = process.system_output("lparstat -E 1 1").decode("utf-8")
+        input_string = ''
+        for line in output.splitlines():
+            if 'GHz' in line:
+                input_string = line
+                # Define the regular expression pattern
+                pattern = r'(\d+\.\d+)\s+(\d+\.\d+)\s+\d+\.\d+GHz\[\s*(\d+)%\]\s+(\d+\.\d+)\s+(\d+\.\d+)'
```
Comment (Naresh-ibm) [inline ras/ras_ppcutils.py:467]: @shirishaganta1 looks like the length is >79. please take care of it.

**Review by Naresh-ibm — CHANGES_REQUESTED** [review]

File: ras/ras_ppcutils.py [hunk 2]
```diff
@@ -447,3 +449,40 @@ def test_lparnumascore(self):
...
+                    normal_idle = float(data[3])
+                    normal_busy = float(data[4])
+                    normal = normal_idle + normal_busy
+                    freq_percentile = float(data[2])
+        if (actual_busy > 0) and (actual_idle < 100):
+            self.log.info(
+                "Actual busy and idle values are displaying  correctly")
+        else:
+            self.cancel(
+                "busy value and idle values are not  displaying correctly")
```
Comment (abdhaleegit) [inline ras/ras_ppcutils.py:482]: should fail instead of cancel ?

File: ras/ras_ppcutils.py [hunk 3]
```diff
@@ -447,3 +449,40 @@ def test_lparnumascore(self):
...
+        if normal == freq_percentile:
+            self.log.info("normal busy and idle are matching with the \
+                           percentage under frequency")
+        else:
+            self.cancel("normal busy and value percentage are not matching \
+                        with the frequency percentile")
```
Comment (abdhaleegit) [inline ras/ras_ppcutils.py:488]: here too should fail instead of cancel ?

**Review by abdhaleegit — CHANGES_REQUESTED** [review]
we can always avoid one condition in if /else  and only log if fail .. and ignore log.info if pass

File: ras/ras_ppcutils.py (hunk 1 above)
Comment (shirishaganta1) [reply ras/ras_ppcutils.py]: sure will take care of that

File: ras/ras_ppcutils.py (hunk 3 above)
Comment (shirishaganta1) [reply ras/ras_ppcutils.py:489]: yeah, for the above two cases I have changed it to fail if the condition is not met

**Review by sacsant — COMMENTED** [review]
Why do we need a separate test_ routine? Can we add this code to existing test_lparstat() routine ?

File: ras/ras_ppcutils.py (hunk 2 above)
Comment (shirishaganta1) [reply ras/ras_ppcutils.py]: > Why do we need a separate test_ routine? Can we add this code to existing test_lparstat() routine ?

yes we can do that..

**Comment by Naresh-ibm** [comment]
@shirishaganta1 please re-push taking care of Travis and above suggested changes

**Comment by shirishaganta1** [comment]
Addressed all the changes suggested above and re-pushed it

**Comment by Naresh-ibm** [comment]
@shirishaganta1 are we still pursuing this  in ras_ppcutils.py?
Or
as suggested by Sachin, will add this code to existing test_lparstat() routine ?

**Review by PraveenPenguin — APPROVED** [review]
LGTM @Naresh-ibm it already moved in lpar stats

File: ras/ras_ppcutils.py [hunk 4]
```diff
@@ -428,6 +433,35 @@ def test_lparstat(self):
...
+                    actual_idle = float(data[1])
+                    normal_idle = float(data[3])
+                    normal_busy = float(data[4])
+                    normal = normal_idle + normal_busy
+                    freq_percentile = float(data[2])
+        if (actual_busy > 0) and (actual_idle < 100):
+            self.log.info(
+                "Actual busy and idle values are displaying  correctly")
+        else:
+            self.fail("busy value and idle values are not \
```
Comment (sacsant) [inline ras/ras_ppcutils.py:457]: Busy and idle actual values are incorrect.

File: ras/ras_ppcutils.py [hunk 5]
```diff
@@ -428,6 +433,35 @@ def test_lparstat(self):
...
+                       displaying correctly")
+        if normal == freq_percentile:
+            self.log.info("normal busy and idle are matching with the \
```
Comment (sacsant) [inline ras/ras_ppcutils.py:460]: Normalised busy plus idle value matches with Frequency (percentage)

File: ras/ras_ppcutils.py [hunk 6]
```diff
@@ -428,6 +433,35 @@ def test_lparstat(self):
...
+                           percentage under frequency")
+        else:
+            self.fail("normal busy and value percentage are not matching \
```
Comment (sacsant) [inline ras/ras_ppcutils.py:463]: Normalised busy plus idle value does not match with Frequency (percentage)

Comment (shirishaganta1) [reply ras/ras_ppcutils.py]: @sacsant As Suggested above changed the sentences and re-pushed it again

**Review by sacsant — COMMENTED** [review]
Thanks. Looks good to me.

File: ras/ras_ppcutils.py [hunk 7]
```diff
@@ -14,8 +14,10 @@
 # Copyright: 2016 IBM
 # Author: Pavithra <pavrampu@linux.vnet.ibm.com>
 # Author: Sachin Sant <sachinp@linux.ibm.com>
+# Author: Shirisha Ganta <shirisha.ganta1@ibm.com>
```
Comment (abdhaleegit) [inline ras/ras_ppcutils.py:17]: A common practice is to use vnet id for Author and also for sign-off,  ignore if the id is vnet id

File: ras/ras_ppcutils.py [hunk 8]
```diff
@@ -428,6 +433,33 @@ def test_lparstat(self):
                 self.log.info("%s command passed" % cmd)
                 self.fail("lparstat: Expected failure, %s command exeucted \
                           successfully." % cmd)
+        output = process.system_output("lparstat -E 1 1").decode("utf-8")
+        input_string = ''
+        for line in output.splitlines():
+            if 'GHz' in line:
+                input_string = line
```
Comment (abdhaleegit) [inline ras/ras_ppcutils.py:440]: cannot I use line directly, any need to assign to a new param ?

**Review by abdhaleegit — APPROVED** [review]
Rest all LGTM

File: ras/ras_ppcutils.py (hunk 7 above)
Comment (shirishaganta1) [reply ras/ras_ppcutils.py:17]: from next Patch onwards will do that as currently I need to do few more steps to get the vnet id..

File: ras/ras_ppcutils.py (hunk 8 above)
Comment (shirishaganta1) [reply ras/ras_ppcutils.py]: yes, we can use the line directly .. Changed the code and re-pushed it again

**Review by Naresh-ibm — APPROVED** [review]
@shirishaganta1 Thanks for all changes and PR
@abdhaleegit @PraveenPenguin @sacsant Thanks for reviews and suggestions.
LGTM

**Pull request merged**